File: Problems/1046. (E) Last Stone Weight/Java/solution.java
class Solution {
  public int lastStoneWeight(int[] stones) {
    // heap queue

    // Java pqueues also use min heaps as default, use max heap instead
    Queue<Integer> hq = new PriorityQueue(Collections.reverseOrder());
    for (int i: stones)
      hq.add(i);

    while (hq.size() > 1) {
      int y = hq.poll();
      int x = hq.poll();
      if (x == y)
        continue;
      hq.add(y-x);
    }
    return hq.size() == 0 ? 0 : hq.poll();
  }
}

File: Problems/1046. (E) Last Stone Weight/Python3/solution.py
class Solution:
  def lastStoneWeight(self, stones: List[int]) -> int:
    # heap queue

    # Python max heap hack
    # yes, there probably is a better way of doing this
    hq = [-1*i for i in stones]
    heapify(hq)

    while len(hq) > 1:
      y = heappop(hq) * -1
      x = heappop(hq) * -1
      if x == y:
        continue
      heappush(hq, -1*(y-x))

    return 0 if not hq else -1*hq[0]

